### DISCUSSION TITLE
Polygons are broken in some cases, can be fixed by adding 0.01 to coordinates.

### DISCUSSION THREAD

**Discussion by georgy-moshkin** [thliebig/openEMS-Project#207] [General]
I can't understand why this is happening. Maybe there is some error in my code, but I do not see it. Here is how problem looks like:
![image](https://github.com/thliebig/openEMS-Project/assets/48346176/2d87b012-e893-4276-91b7-e275d80523e0)
But after replacing pp(2,16) = 0; with pp(2,16) = **0.01;** it looks normal.
`

  for s = 0:patch.N-1
    pp(1,1) = 0;                    pp(2,1) = 0;
    pp(1,2) = 0;                    pp(2,2) = len180/2-feed.width*3/2;

    pp(1,3) = addLen-feed.width;    pp(2,3) = len180/2-feed.width*3/2;
    pp(1,4) = addLen;               pp(2,4) = len180/2-feed.width*3/2+feed.width;

    pp(1,5) = addLen;               pp(2,5) = len180/2-feed.width*3/2+feed.width*3-feed.width;
    pp(1,6) = addLen-feed.width;    pp(2,6) = len180/2-feed.width*3/2+feed.width*3;

    pp(1,7) = 0;                    pp(2,7) = len180/2-feed.width*3/2+feed.width*3;
    pp(1,8) = 0;                    pp(2,8) = len180;

    pp(1,9)  = -feed.width;         pp(2,9)  = len180;

    pp(1,10) = -feed.width;             pp(2,10) = len180/2+feed.width/2+feed.width;
    pp(1,11) = -feed.width+feed.width;  pp(2,11) = len180/2+feed.width/2;

    pp(1,12) = -feed.width+addLen;      pp(2,12) = len180/2+feed.width/2;
    pp(1,13) = -feed.width+addLen;      pp(2,13) = len180/2-feed.width/2;

    pp(1,14) = -feed.width+feed.width;  pp(2,14) = len180/2-feed.width/2;
    pp(1,15) = -feed.width;             pp(2,15) = len180/2-feed.width/2-feed.width;

    pp(1,16) = -feed.width;             pp(2,16) = 0;

    for k=1:16
      pp(1,k)=pp(1,k)+feed.width/2;
      pp(2,k)=pp(2,k)+offsY-len180-patch.length/2;
    endfor

    CSX = AddPolygon( CSX, 'network', 1, 'z', substrate.thickness, pp);

  endfor

`

**Comment by BryceBuschman** [comment]
Hi @georgy-moshkin,

Have you tried to round your coordinates to 5 or 6 decimal places? I had a problem similar to this. In my cased it seemed like decimal truncation when building the polygon caused some points to fall on-top of each other/get too close to each other. Rounding to 6 decimal places fixed the problem. Good luck!

**Reply by Juleinn** [reply]
Had some trouble with this too, when using floating point for defining geometry floating-point errors would accumulate to the point where it would create close points causing cell size to be very small and thus computation time very large. 

Something like
```python 
co = [[round(i[0], 5), 
  round(i[1], 5), 
  round(i[2], 5)] for i in local_verts]
``` 
works for me

Using fixed-point for geometry would be a good improvement I think (like in KiCAD where everything is nanometers as integers), but that's a significant amount of work to do

**Comment by georgy-moshkin** [comment]
Thanks for the suggestions. In my case, I fixed it by adding "pp=0;" in the end of the loop:
```
for n = 0:block.N-1
% some code fills pp(i,k) = ...
CSX = AddPolygon( CSX, 'top_metal', 1, 'z', substrate.thickness, pp);
pp=0; % erase pp() contents - THIS FIXES THE ISSUE
endfor;
```
I think this destroys 2-dimensional pp(i,k) array, so it can be recreated from the scratch on each iteration.
I haven't look into it afterwards and still don't understand what's going on here. In other words, I don't understand why result is being different, because I have fixed number of elements, a 2 x 16 array which is simply being overwritten with new values on each iteration.

**Comment by thliebig** [comment]
I think this is only a glitch in the visualization by AppCSXCAD?  I think this is a bug in QCSXCAD and not in CSXCAD or openEMS. But I would need to confirm that...

**Reply by tmayo-d3** [reply]
It seems to be a bug in visualization (see https://github.com/thliebig/AppCSXCAD/issues/4), but this makes it pretty hard to design and verify a good mesh since there is no way to see what the mesh and design look like.  This is very brutal for real PCB designs where the design is brought in from HyperLynx vs. small hand-coded circuits.  I'm looking in https://github.com/thliebig/QCSXCAD/blob/master/QVTKStructure.cpp, but I don't have things building and that's a big hurdle for me.  Is there any hope in having you look into this Thorsten?

**Reply by tmayo-d3** [reply]
@thliebig I'm tagging you in case this helps you see this message.

**Reply by VolkerMuehlhaus** [reply]
> This is very brutal for real PCB designs where the design is brought in from HyperLynx vs. small hand-coded circuits.

@tmayo-d3 Are you sure your **input** data for creating polygon vertices is clean and polygons are closed properly, with no offset from numercial precision issues?

For my workflow that creates openEMS models from GDSII layouts using python API, I have never seen such issues, and these designs can be really complex.

**Reply by tmayo-d3** [reply]
I used the ImportHyperLynx Matlab function to bring in the board design.  The XML file it generates is definitely the issue.

From this XML file:
Vertex X1="0.05382" X2="0.07469"

These should be:53.82246538 74.68964488

Which came from the HyperLynx file as:
(LINE X=2.1189947 Y=2.9405372)

These numbers times 25.4 is how I get 53.82246538 and 74.68964488.

So, I guess the issue is with ImportHyperLynx somehow.

I tried lowering the grid to 1e-12, and I got:
Vertex X1="0.0538225" X2="0.0746896"

Which seems better, but this still plots with discontinuities in AppCSXCAD.

I updated this because some numbers were blocked by angle braces.

**Comment by tmayo-d3** [comment]
Hyp2Mat seems to use only FLOATs in the parsing of the input file.  I gather this is only around 7 digits of precision, which is what I have.  I'd love to know if there is another workflow to get PCB designs into OpenEMS.

**Reply by VolkerMuehlhaus** [reply]
A user of my gds2openEMS workflow had demonstrated that this can be used for PCB data also, if you use klayout editor to read Gerber and convert to GDSII layout, layer by layer. This is documented at the end of the readme in that repo. https://github.com/EngGhaith/VMH_Flow_4_PCB

However, this workflow is meant for RFIC layout, and abusing it for PCB is somewhat overkill with too much manual work and overhead.

I think it's better to keep searching for the root cause of your problem. @georgy-moshkin had some comment above which might help, and that seems to be related to data handling/variable cleanup in the code (not accuracy).
https://github.com/thliebig/openEMS-Project/discussions/207#discussioncomment-11138765

**Reply by tmayo-d3** [reply]
I do clear out the vertices with p = 0 between polygons.  This doesn't improve anything.

**Reply by tmayo-d3** [reply]
I have a feeling we need DOUBLE instead of FLOAT in parse.yy in hyp2mat/lib.

**Reply by tmayo-d3** [reply]
I've taken the hyp2mat output, and scaled all of the points by a factor of 100, and AppCSXCad now plots things correctly.  I wonder if there is something about the frequency I am using (60 GHz) having such small dimensions in meters that things are not represented well that way.  I'm going to read how OpenEMS handles units a bit more.  Maybe someone has come across this angle of things before.

**Reply by VolkerMuehlhaus** [reply]
@tmayo-d3 I am also doing RFIC simulation with dimensions of a few microns in openEMS, frequency range DC to > 100 GHz. This is all Python API these days, but a few years ago I also created Matlab API example models, which you can find here:
https://github.com/IHP-GmbH/IHP-Open-PDK/tree/main/ihp-sg13g2/libs.tech/openems/obsolete/testcase/SG13_Octagon_L2n0/OpenEMS_Matlab
Geometries are created as extruded polygons with numbers in micron, and unit 1E-6 specified in DefineRectGrid().

**Reply by tmayo-d3** [reply]
Thank you very much for this info @VolkerMuehlhaus . I think my issue stems from the fact that I am importing HyperLynx files.  I need to figure out how I can specify the output units somehow.  It looks like the output units from that are in meters, and there is not enough precision with the FLOATs that are used in parse.yy.  I'm weighing modifying parse.yy vs. writing another tool to process the HyperLynx file first, convert to metric with enough precision, and remove the ENGLISH directive.  We looked and could not find a way to make Altium output in metric.

**Reply by tmayo-d3** [reply]
I've spent all day trying to set up to build hyp2mat.  Any help or advice here on using hyp2mat with boards in mmWave frequencies would be helpful.  Who has solved this before?  I looked at scaling the CSX design, but it looks like that has to be done per component, not the whole design.
